From 91e9d120c308ef3b83bf3c53c5f8bb4d7daf1c41 Mon Sep 17 00:00:00 2001
From: Thang Tran <thuutran@amperecomputing.com>
Date: Tue, 14 Jan 2025 15:38:30 +0700
Subject: [PATCH] transport: check ifaddr before access to it's property

Issue: When user request to switch from dhcp to static IP configuration,
then request to read LAN status, the ipmid is crashed.

Steps:
  1. Change to the static IP configuration
     $ipmitool lan set 1 ipsrc static
  2. Show the LAN information
     $ipmitool lan print 1
  3. ipmid application is crashed

Root cause: ipmid did not check the ifaddr is valid or not before access
to it's property.

Solution: Check the ifaddr is valid or not before access to it's
property.

Upstream-Status: Pending

Tested:
  1. Change to the static IP configuration
     $ipmitool lan set 1 ipsrc static
  2. Show the LAN information
     $ipmitool lan print 1
  3. ipmid does not be crashed.

Signed-off-by: Thang Tran <thuutran@amperecomputing.com>
---
 transporthandler.cpp | 19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

diff --git a/transporthandler.cpp b/transporthandler.cpp
index 772925e..46d9691 100644
--- a/transporthandler.cpp
+++ b/transporthandler.cpp
@@ -271,12 +271,14 @@ auto getIfAddr4(sdbusplus::bus_t& bus, const ChannelParams& params)
     for (uint8_t i = 0; i < MAX_IPV4_ADDRESSES; ++i)
     {
         ifaddr4 = getIfAddr<AF_INET>(bus, params, i, originsV4);
-        if (src == ifaddr4->origin)
-        {
-            break;
-        }
-        else
+
+        if (ifaddr4)
         {
+            if (src == ifaddr4->origin)
+            {
+                return ifaddr4;
+            }
+
             ifaddr4 = std::nullopt;
         }
     }
@@ -301,12 +303,17 @@ void reconfigureIfAddr4(sdbusplus::bus_t& bus, const ChannelParams& params,
         elog<InternalFailure>();
     }
     uint8_t fallbackPrefix = AddrFamily<AF_INET>::defaultPrefix;
+
+    auto addr = stdplus::In4Addr{};
     if (ifaddr)
     {
         fallbackPrefix = ifaddr->prefix;
         deleteObjectIfExists(bus, params.service, ifaddr->path);
+        addr = ifaddr->address;
     }
-    auto addr = address.value_or(ifaddr->address);
+
+    addr = (address != std::nullopt) ? address.value() : addr;
+
     if (addr != stdplus::In4Addr{})
     {
         createIfAddr<AF_INET>(bus, params, addr,
-- 
2.43.0

